src/internal_modules/roc_core/heap_arena.cpp

```diff
@@ -118,5 +118,33 @@ void HeapArena::deallocate(void* ptr) {
     free(chunk);
 }
 
+size_t HeapArena::compute_allocated_size(size_t size) const {
+    return sizeof(ChunkHeader) + sizeof(ChunkCanary) + size + sizeof(ChunkCanary);
+}
+
+size_t HeapArena::allocated_size(void* ptr) const {
+    if (!ptr) {
+        roc_panic("heap arena: null pointer");
+    }
+
+    ChunkHeader* chunk =
+        ROC_CONTAINER_OF((char*)ptr - sizeof(ChunkCanary), ChunkHeader, data);
+
+    const bool is_owner = chunk->owner == this;
+
+    if (!is_owner) {
+        if (AtomicOps::load_seq_cst(flags_) & HeapArenaFlag_EnableGuards) {
```

Comment:
I had to remove `num_guard_failures_++;` from here, because this method is const. Would you rather have the member variable as mutable?

Reply:
It makes sense to make it mutable. I'd say it's not part of the logical object state.